### ISSUE TITLE
`bitvector` should not use `syntax-id-rules`

### ISSUE THREAD

**Issue by stchang** [emina/rosette#26]
Currently, `bitvector` is broken because `syntax-id-rules` is poorly designed (and should not be used):

```racket
> (bitvector 1 1)
#<procedure:bitvector-type>
> (bitvector 1 2 3 4 6 #:fskal)
#<procedure:bitvector-type>
> ((bitvector 1 2 3 4 6 #:fskal) 1)
(bitvector 1)
> ((set! bitvector 5) 1)
(bitvector 1)
```

A better alternative is to ([in this code](https://github.com/emina/rosette/blob/master/rosette/base/core/bitvector.rkt#L70-L75)) use `make-variable-like-transformer`:

```racket
(define-match-expander @bitvector
  (syntax-rules ()
    [(_ sz) (bitvector sz)])
  (make-variable-like-transformer #'bitvector-type))
```

Should we submit a pull request with this change?

**Comment by emina** [comment]
That would be great.  Thanks!

**Comment by AlexKnauth** [comment]
I'll make a pull request to change it, but would you rather this depended on Racket version 6.3, or on the unstable-lib package?

**Comment by emina** [comment]
I'll be updating Rosette to Racket v6.6 very soon.  So, if it will work with v6.6, that would be perfect.

**Comment by AlexKnauth** [comment]
If I use `unstable/struct` it will work in versions both before and after 6.3, but require unstable-lib. If I use `syntax/transformer` it will work only after 6.3. Both will work with 6.6, but if someone with 6.1 or 6.2 wanted to use rosette, I would need to use `unstable/struct`.

**Comment by emina** [comment]
Ah, Rosette users will need to update to Racket v6.6 anyway, due to the changes I will be pushing soon.  These take advantage of the new features in #lang info.

**Comment by AlexKnauth** [comment]
What new features?

**Comment by emina** [comment]
This:

```(define test-omit-paths (if (getenv "PLT_PKG_BUILD_SERVICE") 'all '()))```

**Comment by AlexKnauth** [comment]
Oh. I guess I'll use `syntax/transformer` then.

**Comment by emina** [comment]
Yep, that would be great!

**Comment by AlexKnauth** [comment]
Context: [this code here](https://github.com/emina/rosette/blob/master/rosette/query/debug.rkt#L44-L52)

In the definitions of `true` and `false` in `rosette/query/debug.rkt`, are `(true . whatever)` and `(false . whatever)` supposed to be equivalent to using `true` and `false` as identifiers? Also, what is `relax` supposed to do?

I'm asking because I'm wondering if I can translate them to `(make-variable-like-transformer #'(relax #t true))`, or if should do something more complicated, with `syntax-case`. How were these `true` and `false` definitions meant to be used?

**Comment by emina** [comment]
They are meant to be used as the identifiers true and false.  The `relax` procedure is necessary to allow debugging of incorrect boolean constants.  Run [sdsl/fsm/demo.rkt](https://github.com/emina/rosette/blob/master/sdsl/fsm/demo.rkt) to see an example.

**Comment by AlexKnauth** [comment]
So is `relax` a function, or is it a macro? Should `(true . whatever)` be an error? Or should it expand to `(relax #t (true . whatever))`, or `(relax #t true)`, or `(relax #t old-true)`, or?

If I use `true` with the current `syntax-id-rules` definition, it loops forever.

**Comment by AlexKnauth** [comment]
And if I try using `true` in a `#lang rosette` file, it errors with `relax: unbound identifier in module`. 

Should it be using `relax?`, or `relax-values`, or `relaxer`, or `relaxer?`, or?

**Comment by emina** [comment]
Ah, this is leftover code from Rosette 1.0 that I forgot to remove from 2.x.  Removed now.

**Comment by AlexKnauth** [comment]
Context: [this code here](https://github.com/emina/rosette/blob/master/sdsl/ifc/value.rkt#L75-L80)

When `R` is used as an identifier, should the function it returns have an optional argument just like the macro does?

**Comment by stchang** [comment]
In general, should we try to fix other possibly buggy uses of `syntax-id-rules`? e.g., a quick search shows that there might be a problem here as well: https://github.com/emina/rosette/blob/c37b507554d70618053b5605a7b9f37e69ba1e09/rosette/solver/solution.rkt#L56-L70

**Comment by emina** [comment]
Let's just deal with the original issue, and leave other uses of syntax-id-rules as-is.

**Comment by stchang** [comment]
Ok sounds good. Alex, do you want to submit a pull request for just `bitvector`?

**Comment by AlexKnauth** [comment]
No, I've already made https://github.com/emina/rosette/pull/27

**Comment by emina** [comment]
Alex, I'm not comfortable with such an invasive and extensive change (#27), so I will close that pull request.  I'm happy to merge a PR just for `bitvector` though.

**Comment by stchang** [comment]
@AlexKnauth, just fix the `bitvector` issue.
@emina, if we find other misbehaving functions, should we submit a separate issue ticket?

**Comment by emina** [comment]
@stchang Yes, that would be great.  I definitely want to know about / merge anything that is blocking your progress.  I take it the `bitvector` issue came up when trying to expand typed Rosette into untyped?

**Comment by stchang** [comment]
Actually, our typed Rosette prevents the bad uses of `bitvector` :) I only discovered the bug by accident while writing my own test cases.

I looked at Alex's #27 and it does seem like it fixes many other functions that are exhibiting the same problem, e.g.:

```racket
$ racket
Welcome to Racket v6.6.0.2.
-> (require rosette)
-> (bv 1 2 3 4 5 6 7)
#<procedure:make-bv>
-> (model 1 2 3 4 5 6)
#<procedure:get-model>
```

Though it makes sense to first make a small change and see how much user code is affected.

**Comment by emina** [comment]
Cool!  Yeah, I prefer conservative changes when it comes to code that's out in the wild :)
